llama-cli/src/main.rs

```diff
@@ -91,15 +94,53 @@ fn main() {
                     );
                 }
             }
-        })
-        .expect("Could not load model");
+        },
+    )
+    .expect("Could not load model");
 
     log::info!("Model fully loaded!");
 
     let mut rng = thread_rng();
-    model.inference_with_prompt(&vocab, &inference_params, &prompt, &mut rng, |t| {
-        print!("{t}");
-        std::io::stdout().flush().unwrap();
-    });
-    println!();
+
+    if let Some(restore_path) = &args.restore_prompt {
+        let snapshot = InferenceSnapshot::load_from_disk(restore_path);
+        match snapshot.and_then(|snapshot| model.set_snapshot(snapshot)) {
+            Ok(_) => {
+                log::info!("Restored cached memory from {restore_path}");
+            }
+            Err(err) => {
+                eprintln!("Could not restore prompt. Error: {err}");
+                std::process::exit(1);
+            }
+        }
+    }
+
+    if let Some(cache_path) = &args.cache_prompt {
+        model.feed_prompt(&vocab, &inference_params, &prompt, |t| {
+            print!("{t}");
+            std::io::stdout().flush().unwrap();
+        });
+        println!();
+
+        // Write the memory to the cache file
+        // SAFETY: no other model functions used inside the block
+        unsafe {
+            let memory = model.get_snapshot();
```

Comment:
Is there a way to make this safe? Something like `memory.with_snapshot(|snapshot| {})` so that access is bounded

Reply:
Basically, this "unsafe" is here out of caution. This function should not be marked unsafe, because the returned object has the correct lifetime and prevents mutating the model for as long as it lives.

The problem here is that tensors inside `ggml` can be mutated through a shared ref `&`. When writing the function to return the slice I realized all this and thought: "Let's just mark this as unsafe for the time being until we figure a better way".

The whole "safety-ness" of the library needs to be reviewed, but I'd do that in another PR. For now, let's just push the requirements to the caller on this one.

Reply:
Fair enough, let's revisit later!